tools/workload-simulator.html

```diff
@@ -0,0 +1,1160 @@
+          minFilter: 'linear',
+        });
+        const texture = device.createTexture({
+          size: [256, 256, 1],
+          format: 'rgba8unorm',
+          usage: GPUTextureUsage.SAMPLED | GPUTextureUsage.COPY_DST,
+        });
+        const loadImage = async () => {
+          const imageBitmap = await createImageBitmap(img);
+          device.queue.copyImageBitmapToTexture(
+            { imageBitmap }, { texture }, [256, 256, 1]);
+          render();
+        };
+        if (img.complete) loadImage(); else img.addEventListener('load', loadImage);
+        uniformBuffer = device.createBuffer({
+          size: 2 * 4,
+          usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
+        });
+        bindGroup = device.createBindGroup({
+          layout: bindGroupLayout,
+          entries: [
+            { binding: 0, resource: { buffer: uniformBuffer } },
+            { binding: 1, resource: sampler },
+            { binding: 2, resource: texture.createView() },
+          ],
+        });
+      } catch(e) {
+        errorMessage.textContent = "Error initializing WebGPU: " + e;
+        return;
+      }
+    }
+
+    // End WebGPU init.
+    // Begin WebGPU rendering.
+
+    if (!device) {
+      errorMessage.textContent = "Error: WebGPU not supported.";
+      return;
+    }
+    try {
+      const swapChainView = swapChain.getCurrentTexture().createView();
+      const commandBuffers = [];
+      device.queue.writeBuffer(uniformBuffer, 0, new Float32Array([position[0]/size*2 - 0.5, -position[1]*2/size + 0.5]));
+      let buffer = null;
+
+      // Upload data using createBuffer with mappedAtCreation.
+      if (bufferData.value > 0) {
+        buffer = device.createBuffer({ size: bufferDataArray.byteLength, usage: GPUBufferUsage.VERTEX, mappedAtCreation: true });
+        new Float32Array(buffer.getMappedRange()).set(bufferDataArray);
+        buffer.unmap();
+        // TODO: use this data in rendering so it can't be optimized away
+      }
+
+      // Upload data using queue.writeBuffer.
+      if (bufferSubData.value > 0) {
+        if (!uploadBuffer || uploadBufferSize != bufferSubDataArray.byteLength) {
+          if (uploadBuffer) uploadBuffer.destroy();
+          uploadBufferSize = bufferSubDataArray.byteLength;
+          uploadBuffer = device.createBuffer({
+            size: bufferSubDataArray.byteLength,
+            usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
+          });
+        }
+        device.queue.writeBuffer(uploadBuffer, 0, bufferSubDataArray);
+      }
+
+      // Upload data using mapAsync and a queue of staging buffers.
+      if (mapAsyncArray.length > 0) {
+        if (mapAsyncBufferSize != mapAsyncArray.byteLength) {
+          mapAsyncBufferSize = mapAsyncArray.byteLength;
+          if (mapAsyncBuffer) mapAsyncBuffer.destroy();
+          mapAsyncBuffer = device.createBuffer({size: mapAsyncArray.byteLength, usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST});
```

Comment:
why is this called `mapAsyncBuffer`? it's not mappable

Reply:
It's the ultimate destination buffer for the mapAsync upload path. I'll rename it to mapAsyncTargetBuffer.